core/src/cluster_slot_state_verifier.rs

```diff
@@ -255,76 +332,546 @@ pub(crate) fn check_slot_agrees_with_cluster(
+                ResultingStateChange::RepairDuplicateConfirmedVersion(cluster_duplicate_confirmed_hash.unwrap())],
+            )
+        },
+        duplicate_confirmed_state_update_0: {
+            // Duplicate confirmed state has to be Some
+            let cluster_duplicate_confirmed_hash = Some(Hash::new_unique());
+            let bank_frozen_hash = Hash::default();
+            (
+                SlotStateHandlerArgs {
+                    slot: 1,
+                    bank_frozen_hash,
+                    cluster_duplicate_confirmed_hash,
+                    is_dead: false,
+                    is_slot_duplicate: false,
+                },
+                SlotStateUpdate::DuplicateConfirmed,
+                Vec::<ResultingStateChange>::new()
+            )
+        },
+        duplicate_confirmed_state_update_1: {
+            // Duplicate confirmed state has to be Some
+            let cluster_duplicate_confirmed_hash = Some(Hash::new_unique());
+            let bank_frozen_hash = Hash::default();
+            (
+                SlotStateHandlerArgs {
+                    slot: 1,
+                    bank_frozen_hash,
+                    cluster_duplicate_confirmed_hash,
+                    is_dead: false,
+                    is_slot_duplicate: true,
+                },
+                SlotStateUpdate::DuplicateConfirmed,
+                Vec::<ResultingStateChange>::new()
+            )
+        },
+        duplicate_confirmed_state_update_2: {
+            // Duplicate confirmed state has to be Some
+            let cluster_duplicate_confirmed_hash = Some(Hash::new_unique());
+            let bank_frozen_hash = cluster_duplicate_confirmed_hash.unwrap();
+            (
+                SlotStateHandlerArgs {
+                    slot: 1,
+                    bank_frozen_hash,
+                    cluster_duplicate_confirmed_hash,
+                    is_dead: false,
+                    is_slot_duplicate: false,
+                },
+                SlotStateUpdate::DuplicateConfirmed,
+                vec![
+                ResultingStateChange::DuplicateConfirmedSlotMatchesCluster(bank_frozen_hash)]
+            )
+        },
+        duplicate_confirmed_state_update_3: {
+            // Duplicate confirmed state has to be Some
+            let cluster_duplicate_confirmed_hash = Some(Hash::new_unique());
+            let bank_frozen_hash = cluster_duplicate_confirmed_hash.unwrap();
+            (
+                SlotStateHandlerArgs {
+                    slot: 1,
+                    bank_frozen_hash,
+                    cluster_duplicate_confirmed_hash,
+                    is_dead: true,
+                    is_slot_duplicate: true,
+                },
+                SlotStateUpdate::DuplicateConfirmed,
+                vec![
+                ResultingStateChange::DuplicateConfirmedSlotMatchesCluster(bank_frozen_hash)]
+            )
+        },
+        duplicate_confirmed_state_update_4: {
```

Comment:
Do we want to check case 4 & 5 with the slot being dead as well? I don't think we filter out dead frozen slots on DuplicateConfirmed right?

Reply:
In cases 4 and 5 the `bank_frozen_hash` is not default, so the slot cannot be dead. `duplicate_confirmed_state_update_6` and `duplicate_confirmed_state_update_7` test when the bank .Hash is default and dead